### ISSUE TITLE
ActivityNotFoundException (No browser installed on device)

### ISSUE THREAD

**Issue by esperaking81** [openid/AppAuth-Android#521]
How to handle the case where no browser is installed on the device ? Can't we launch the authorization request from an webview ?

**Comment by chiara-jm** [comment]
we are facing something similar:

`Caused by android.content.ActivityNotFoundException
       at net.openid.appauth.AuthorizationService.prepareAuthorizationRequestIntent + 372(AuthorizationService.java:372)
       at net.openid.appauth.AuthorizationService.getAuthorizationRequestIntent + 267(AuthorizationService.java:267)
       at com.volvocars.openid.SignInActivity.signIn + 62(SignInActivity.java:62)
       at com.volvocars.openid.SignInActivity.onCreate + 36(SignInActivity.java:36)
       at android.app.Activity.performCreate + 7183(Activity.java:7183)
       at android.app.Instrumentation.callActivityOnCreate + 1220(Instrumentation.java:1220)
       at android.app.ActivityThread.performLaunchActivity + 2910(ActivityThread.java:2910)
       at android.app.ActivityThread.handleLaunchActivity + 3032(ActivityThread.java:3032)
       at android.app.ActivityThread.-wrap11(ActivityThread.java)
       at android.app.ActivityThread$H.handleMessage + 1696(ActivityThread.java:1696)
       at android.os.Handler.dispatchMessage + 105(Handler.java:105)
       at android.os.Looper.loop + 164(Looper.java:164)
       at android.app.ActivityThread.main + 6944(ActivityThread.java:6944)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.Zygote$MethodAndArgsCaller.run + 327(Zygote.java:327)
       at com.android.internal.os.ZygoteInit.main + 1374(ZygoteInit.java:1374)`

**Devices**
samsung 45%
Sony 31%
HMD Global 10%
bq 9%
Other… 5%

**Operating Systems**
9 - 43%
10 - 39%
7 - 11%
8 - 7%

**Comment by kiskunk** [comment]
How can we get a proper callback for this? We need to show an error message to the user to tell them to install some browsers, but `getAuthorizationRequestIntent` doesn't return forever if there is no activity to handle it.

**Comment by Barryrowe** [comment]
@kiskunk You should be able to do something like:

```
try  {
  val authIntent =  authService. getAuthorizationRequestIntent(authRequest)
  myMethodToStartAuthFlow(authIntent)
}catch(err) {
  when (err) {
     is ActivityNotFoundException -> showBrowserRequiredMessage()
     else -> showGenericError()
}
```

We were surprised by the number of  devices encountering this as well. We now catch this and present the user a message letting them know they require a browser to continue.

**Comment by iainmcgin** [comment]
It sounds like throwing a formally documented exception on calls that require browsers would be helpful here. No guarantees can be provided as to when this can be implemented, but it's a reasonable request.

**Comment by amirboys07** [comment]
difficult

**Comment by jannisveerkamp** [comment]
We also see this problem in our logs. We actually have one Testdevice (Samsung S7) where this error occured for whatever reasons.

**Setup:** Chrome (default), Firefox and Opera are installed.

We found that this line in `BrowserSelector.java`:
```
List<ResolveInfo> resolvedActivityList =
                pm.queryIntentActivities(BROWSER_INTENT, queryFlag);
```
resolves **only for the Opera Browser** which let everything fail

Note: I have another S7 with the same setup where this error does not occur 🤷‍♀️

**Comment by Eightyplus** [comment]
We experience the same problem. I tried to solve this with various approaches, but I have not really found any good way to handle this. (right now using `try/catch`)

Would be nice with more insight or this error solved
![Screenshot 2020-09-07 at 09 43 38](https://user-images.githubusercontent.com/15194982/92361466-b05d2b00-f0ee-11ea-8d00-49a2a4edc703.png)


According to Firebase we have crashes 
- 67% samsung
- 15% Sony
- 8% asus
- 8% LGE
- 2% Lenovo

Android versions: 
- 61% Android 8
- 26% Android 7
- 10% Android 10
- 3% Android 6

**Comment by MHV1** [comment]
We've been observing this crash on few of our Android users in production. I have managed to reproduce the crash with the sample project in this repo, but only if I choose the "Authorize Auth0" option. There are no browsers installed on the device (Android 10).

We've been using version 5.0.0 of this library, since this seemed to have been handled already: https://github.com/FormidableLabs/react-native-app-auth/pull/443 But it keeps happening. I don't seem to be able to get the error mentioned in the README (browser_not_found (Android only) - no suitable browser installed), only a crash.

I cannot tell for sure if the users experiencing this have no browsers installed or they have an non supported one.

**Comment by pablocoberly** [comment]
We've had this issue on a few Android devices that had installed some other browser, in particular "Photon Browser". It was being used by users to run Flash apps. It would cause no browsers to be returned even if Chrome was installed and set as the default browser. Uninstalling "Photon Browser" solved the problem.

**Comment by SuddenH4X** [comment]
We've experienced the same bug a while ago. If the Opera browser (or Opera beta) is set as default browser a login isn't possible at all. As @jannisveerkamp said the following intent only resolves the Opera browser which isn't able to handle it:

```java
List<ResolveInfo> resolvedActivityList = pm.queryIntentActivities(BROWSER_INTENT, queryFlag);
```
Other browsers installed on the same device don't have any impact on that.

We've tested it on a lot of different devices and found out that the bug only occurs on devices running Android 7 or below. We've also tested about 50 different browsers and couldn't reproduce that strange behavior with another browser.

We've added a workaround for that bug in our fork. The important parts occur in this two commits:
https://github.com/dm-Mobile-SE-Developer/AppAuth-Android/commit/d8786fee1500f1f9da63ef8d19730a57f789b8cc
https://github.com/dm-Mobile-SE-Developer/AppAuth-Android/commit/e2d98fc3e64fbfb75e39d807a6a317a738fc879d

But of course we would be very interested in a fix or workaround in the official repository.

**Comment by carstenhag** [comment]
We have gotten a customer review from a Samsung Galaxy S10+ Android 11 device, and it sounds similar to what @pablocoberly wrote. Unfortunately I can't reproduce it with my Samsung A70 Android 11 test device :(.

**Comment by carstenhag** [comment]
We had answered to the user's review:
>Many thanks for your feedback! We are very sorry you were not able to use the login as intended. We have analyzed the problem. Could you please try to uninstall all browsers except for Chrome? The issue could be caused by Opera or the Photon Browser. After these are removed, the issue should be gone. If this did not solve the issue, please write our support at ....

Unfortunately it did not solve the issue, but the user also did not contact the support, so we are still unsure how to proceed :( The user added this to their review:

>I've followed the developer's advice by uninstalling other EU browsers, it doesn't work any better.

**Comment by agologan** [comment]
@carstenhag is it possible your customer is using an MDM solution and the app is installed in the sandbox?
If so it's quite possible their IT didn't install a browser in their Secure Folder.

**Comment by dsuresh-ap** [comment]
I was able to reproduce this issue on an Android 7 emulator with Opera, Chrome and the WebView browser installed. What I found was that once Opera is set as a default browser, both `ResolveInfo resolvedDefaultActivity =
                pm.resolveActivity(BROWSER_INTENT, 0);` and `List<ResolveInfo> resolvedActivityList = pm.queryIntentActivities(BROWSER_INTENT, queryFlag);` only return the Opera browser. Even after trying the following Opera is the only browser returned:

1. Setting another browser as default
2. Restarting the device
3. Clearing Opera's app data

The only things that worked were:

1. Setting the default browser to another and performing "Reset app preferences" (see attached image)
2. Uninstalling Opera
![image](https://user-images.githubusercontent.com/21132682/121202060-00664b00-c843-11eb-8367-085544264064.png)

The workaround mentioned above by @SuddenH4X worked so I guess we just have to fork the repo and go ahead with that as we have a user with this issue.

**Comment by hugomillwood** [comment]
Hi @agologan,

We're seeing this issue for a few of our customers in production. How would using a BrowserBlacklist for Opera change the fact that the OS only resolves one browser? Our customers are saying that they have other browsers installed, just as some have already mentioned in this issue, but are still not able to login.

Is there a possibility of merging the commits by @SuddenH4X?

**Comment by agologan** [comment]
I have to admit I misunderstood the issue described by @SuddenH4X until I saw and tested the scenario described by @dsuresh-ap 
d4fde4462126fa0f5c183509dc9775d189d8e11b is included in release `0.9.1` so please test and let me know if you notice any further issues.
Testing the exact setup described by @dsuresh-ap on Android 7 no longer returns just one an incorrect Opera entry.

Iain was planning to add maybe a more specific exception to further help with the original OPs issue.
All APIs now document throwing `ActivityNotFoundException` and so I'd consider the original purpose of this issue resolved, especially since it went into a tangent which should also be resolved now.

**Comment by dsuresh-ap** [comment]
@agologan Tested 0.9.1 and I can confirm the fix works. Thanks!

**Comment by bazik123** [comment]
i keep getting the same crash even with 0.9.1

**Comment by agologan** [comment]
@bazik123 as explained in https://github.com/openid/AppAuth-Android/issues/521#issuecomment-872216720 the fix resolves a tangential issue explained by other participants to this issue.

It is expected to receive `ActivityNotFoundException` on the documented APIs if a suitable browser is not available.
If you believe that a suitable browser is actually available on the user's device and the code throws an incorrect `ActivityNotFoundException` then let us know by creating a new issue and providing appropriate contextual information.

**Comment by IgorGanapolsky** [comment]
> @kiskunk You should be able to do something like:
> 
> ```
> try  {
>   val authIntent =  authService. getAuthorizationRequestIntent(authRequest)
>   myMethodToStartAuthFlow(authIntent)
> }catch(err) {
>   when (err) {
>      is ActivityNotFoundException -> showBrowserRequiredMessage()
>      else -> showGenericError()
> }
> ```
> 
> We were surprised by the number of devices encountering this as well. We now catch this and present the user a message letting them know they require a browser to continue.

How come I have Google Chrome, and still get this error??

**Comment by agologan** [comment]
@IgorGanapolsky debugging the issue to provide context or providing a reproducible demo would be helpful.

**Comment by boylenssen** [comment]
@IgorGanapolsky had the same issue in flutter_appauth 1.1.0, but seems to be fixed in flutter_appauth 1.1.0+2

**Comment by jyot103095** [comment]
@IgorGanapolsky would you mind sharing what falls under a suitable browser? Any particular browsers that we can ask our users to avoid ahead of time?

**Comment by ShridharBanavasi1997** [comment]
To provide better accountability for access to installed apps on a device, apps targeting Android 11 (API level 30) will see a filtered list of installed apps by default. In order to access a broader list of installed apps, an app can specify information about apps they need to query and interact with directly. This can be done by adding a element in the Android manifest.

I am able to resolve it by adding
`<Manifest> 
.. 
   <queries> 
        <intent> 
               <action android:name="android.intent.action.VIEW" />
               <category android:name="android.intent.category.BROWSABLE" /> 
               <data android:scheme="https" /> 
        </intent> 
   </queries> 
</Manifest>`

**Comment by agologan** [comment]
@ShridharBanavasi1997 queries is being declared and merged in by the lib since **0.8.0**.
@jairrab Please open a new issue providing proper context. Make sure your users are not using your app in MDM (workplace sandbox) or if they do, that Chrome is installed in that profile.

_Locking this conversation. Feel free to open a new issue referencing this one providing further context if the issue persists.
Use **dsuresh-ap**'s contribution as an example of what can be a helpful in investigating an issue._
